File: WallTeleporters/src/unwrittenfun/minecraft/wallteleporters/blocks/tileentities/TileEntityWallTeleporter.java
package unwrittenfun.minecraft.wallteleporters.blocks.tileentities;

import cpw.mods.fml.common.FMLCommonHandler;
import dan200.computer.api.IComputerAccess;
import dan200.computer.api.ILuaContext;
import dan200.computer.api.IPeripheral;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraftforge.common.ForgeDirection;
import unwrittenfun.minecraft.wallteleporters.blocks.multiblocks.MultiblockWallTeleporter;
import unwrittenfun.minecraft.wallteleporters.handlers.PacketHandler;
import unwrittenfun.minecraft.wallteleporters.info.BlockInfo;

/**
 * Mod: Wall Teleporters
 * Author: UnwrittenFun
 * License: Minecraft Mod Public License (Version 1.0.1)
 */
public class TileEntityWallTeleporter extends TileEntity implements IInventory, IPeripheral {
    public  MultiblockWallTeleporter multiblock = new MultiblockWallTeleporter(this);
    public  int[]                    mask       = new int[]{ 0, 0 };
    private boolean                  loaded     = false;

    public void notifyNeighboursOfConnectionChanged() {
        for (ForgeDirection direction : ForgeDirection.VALID_DIRECTIONS) {
            TileEntity tileEntity = worldObj.getBlockTileEntity(xCoord + direction.offsetX, yCoord + direction.offsetY,
                    zCoord + direction.offsetZ);

            if (tileEntity instanceof TileEntityWallTeleporter) {
                ((TileEntityWallTeleporter) tileEntity).onConnectionChangedFromDirection(direction);
            }
        }
    }

    public void onConnectionChangedFromDirection(ForgeDirection direction) {
        ForgeDirection fromDirection = direction.getOpposite();

        TileEntity tileEntity = worldObj.getBlockTileEntity(xCoord + fromDirection.offsetX,
                yCoord + fromDirection.offsetY, zCoord + fromDirection.offsetZ);

        if (tileEntity == null || tileEntity.isInvalid()) { // Block was removed.
            multiblock.recalculate();
        } else { // Block was added
            if (tileEntity instanceof TileEntityWallTeleporter) {
                TileEntityWallTeleporter teleporter = (TileEntityWallTeleporter) tileEntity;

                MultiblockWallTeleporter.merge(multiblock, teleporter.multiblock);
            }
        }
    }

    public void setMask(int id, int meta) {
        if ((!worldObj.isRemote && !multiblock.isLocked()) || worldObj.isRemote) {
            if (id == BlockInfo.WT_ID) {
                id = 0;
            }
            mask[0] = id;
            mask[1] = meta;

            if (!worldObj.isRemote) {
                PacketHandler.sendNewMaskPacket((byte) 0, this, null);
            }
        }
    }

    public void teleportPlayer(EntityPlayerMP player) {
        if (multiblock.hasDestination()) {
            multiblock.teleportToDestination(player);
        }
    }

    @Override
    public void writeToNBT(NBTTagCompound compound) {
        super.writeToNBT(compound);

        compound.setInteger("MaskId", mask[0]);
        compound.setInteger("MaskMeta", mask[1]);

        if (multiblock.isController(this)) {
            multiblock.writeToNBT(compound);
        }
    }

    @Override
    public void readFromNBT(NBTTagCompound compound) {
        super.readFromNBT(compound);

        mask[0] = compound.getInteger("MaskId");
        mask[1] = compound.getInteger("MaskMeta");

        if (compound.hasKey("WTMultiblock")) {
            new MultiblockWallTeleporter(this);
            multiblock.readFromNBT(compound);
        }
    }

    private int     countdown      = 10;
    private boolean startCountdown = false;

    @Override
    public void updateEntity() {
        if (!loaded && hasWorldObj()) {
            loaded = true;

            if (!worldObj.isRemote && multiblock != null) {
                multiblock.init(worldObj);
            } else if (worldObj.isRemote) {
                notifyNeighboursOfConnectionChanged();
                startCountdown = true;
            }
        }

        if (countdown != -1 && startCountdown) {
            countdown--;
        }

        if (countdown == 0 && multiblock.isController(this)) {
            PacketHandler.requestMultiblockInfoPacket(this);
        }
    }

    @Override
    public void validate() {
        super.validate();

        if (FMLCommonHandler.instance().getEffectiveSide().isClient()) {
            PacketHandler.requestMaskPacket((byte) 0, this);
        }
    }

    @Override
    public void invalidate() {
        super.invalidate();

        notifyNeighboursOfConnectionChanged();
    }


    @Override
    public int getSizeInventory() {
        return multiblock.getSizeInventory();
    }

    @Override
    public ItemStack getStackInSlot(int i) {
        return multiblock.getStackInSlot(i);
    }

    @Override
    public ItemStack decrStackSize(int i, int j) {
        return multiblock.decrStackSize(i, j);
    }

    @Override
    public ItemStack getStackInSlotOnClosing(int i) {
        return multiblock.getStackInSlotOnClosing(i);
    }

    @Override
    public void setInventorySlotContents(int i, ItemStack stack) {
        multiblock.setInventorySlotContents(i, stack);
    }

    @Override
    public String getInvName() {
        return multiblock.getInvName();
    }

    @Override
    public boolean isInvNameLocalized() {
        return multiblock.isInvNameLocalized();
    }

    @Override
    public int getInventoryStackLimit() {
        return multiblock.getInventoryStackLimit();
    }

    @Override
    public boolean isUseableByPlayer(EntityPlayer player) {
        return multiblock.isUseableByPlayer(player);
    }

    @Override
    public void openChest() {
    }

    @Override
    public void closeChest() {
    }

    @Override
    public boolean isItemValidForSlot(int i, ItemStack stack) {
        return multiblock.isItemValidForSlot(i, stack);
    }

    // IPeripheral Implementation

    @Override
    public String getType() {
        return "wallTeleporter";
    }

    @Override
    public String[] getMethodNames() {
        return new String[]{ "getX", "getY", "getZ", "getRotation", "getWorldId", "getWorldName", "clear",
                             "hasDestination", "setUseRotation", "getUseRotation", "setMaskLocked", "isMaskLocked",
                             "setMask", "getMask" };
    }

    @Override
    public Object[] callMethod(IComputerAccess computer, ILuaContext context, int method, Object[] arguments)
            throws Exception {
        switch (method) {
            case 0: // getX
                return new Object[]{ multiblock.destinationX };
            case 1: // getY
                return new Object[]{ multiblock.destinationY };
            case 2: // getZ
                return new Object[]{ multiblock.destinationZ };
            case 3: // getRotation
                return new Object[]{ multiblock.destinationRotation };
            case 4: // getWorldId
                return new Object[]{ multiblock.destinationWorldId };
            case 5: // getWorldName
                return new Object[]{ multiblock.destinationWorldName };
            case 6: // clear
                multiblock.clearDestination();
                break;
            case 7: // hasDestination
                return new Object[]{ multiblock.hasDestination() };
            case 8: // setUseRotation(boolean useRotation)
                if (!(arguments[0] instanceof Boolean)) {
                    throw new Exception("Argument 1 expected to be of type boolean");
                }
                multiblock.setShouldUseRotation((Boolean) arguments[0]);
                break;
            case 9: // getUseRotation
                return new Object[]{ multiblock.shouldUseRotation() };
            case 10: // setMaskLocked(boolean locked)
                if (!(arguments[0] instanceof Boolean)) {
                    throw new Exception("Argument 1 expected to be of type boolean");
                }
                multiblock.setLocked((Boolean) arguments[0]);
                break;
            case 11: // isMaskLocked
                return new Object[]{ multiblock.isLocked() };
            case 12: // setMask(number id, number meta)
                if (!(arguments[0] instanceof Double)) throw new Exception("Argument 1 expected to be of type number");
                if (!(arguments[1] instanceof Double)) throw new Exception("Argument 2 expected to be of type number");
                setMask(((Double) arguments[0]).intValue(), ((Double) arguments[1]).intValue());
                break;
            case 13: // getMask
                return new Object[]{ mask[0], mask[1] };
        }
        return new Object[0];
    }

    @Override
    public boolean canAttachToSide(int side) {
        return true;
    }

    @Override
    public void attach(IComputerAccess computer) {

    }

    @Override
    public void detach(IComputerAccess computer) {

    }
}

File: src/main/scala/unwrittenfun/minecraft/unwrittenblocks/blocks/tileentities/TileEntityWallTeleporter.scala
package unwrittenfun.minecraft.unwrittenblocks.blocks.tileentities

import net.minecraft.inventory.IInventory
import net.minecraft.tileentity.TileEntity
import net.minecraftforge.common.ForgeDirection
import cpw.mods.fml.common.FMLCommonHandler
import unwrittenfun.minecraft.unwrittenblocks.blocks.UBBlocks
import unwrittenfun.minecraft.unwrittenblocks.blocks.multiblocks.MultiblockWallTeleporter
import unwrittenfun.minecraft.unwrittenblocks.handlers.PacketHandler
import net.minecraft.entity.player.{EntityPlayer, EntityPlayerMP}
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.item.ItemStack
import unwrittenfun.minecraft.unwrittenblocks.helpers.CCHelpers
import unwrittenfun.minecraft.unwrittenblocks.network.PacketReceiver
import dan200.computer.api.{ILuaContext, IComputerAccess, IPeripheral}
import scala.Array

/**
 * Mod: UnwrittenBlocks
 * Author: UnwrittenFun
 * License: Minecraft Mod Public License (Version 1.0.1)
 */
class TileEntityWallTeleporter extends TileEntity with IInventory with PacketReceiver with IPeripheral {
  var multiblock: MultiblockWallTeleporter = new MultiblockWallTeleporter() add this
  var mask = Array[Int](0, 0)
  private var _loaded = false
  private var _countdown = 10
  private var _startCountdown = false

  def notifyNeighboursOfConnectionChanged() {
    for (direction <- ForgeDirection.VALID_DIRECTIONS) {
      worldObj.getBlockTileEntity(xCoord + direction.offsetX, yCoord + direction.offsetY, zCoord + direction.offsetZ) match {
        case teleporter: TileEntityWallTeleporter =>
          teleporter onConnectionChangedFromDirection direction
        case _ =>
      }
    }
  }

  def onConnectionChangedFromDirection(direction: ForgeDirection) {
    val opposite: ForgeDirection = direction.getOpposite

    worldObj.getBlockTileEntity(xCoord + opposite.offsetX, yCoord + opposite.offsetY, zCoord + opposite.offsetZ) match {
      case teleporter: TileEntityWallTeleporter =>
        MultiblockWallTeleporter.merge(multiblock, teleporter.multiblock)
      case _ => multiblock.recalculate()
    }
  }

  def setMask(id: Int, meta: Int) {
    if ((!worldObj.isRemote && !multiblock.locked) || worldObj.isRemote) {
      mask(0) = if (id == UBBlocks.wallTeleporter.blockID) 0 else id
      mask(1) = meta

      if (!worldObj.isRemote) PacketHandler.sendNewMaskPacket(0, this, null)
    }
  }

  def teleportPlayer(player: EntityPlayerMP) {
    if (multiblock.hasDestination) multiblock teleportToDestination player
  }

  override def writeToNBT(compound: NBTTagCompound) {
    super.writeToNBT(compound)
    compound.setInteger("MaskId", mask(0))
    compound.setInteger("MaskMeta", mask(1))
    if (multiblock isController this) multiblock writeToNBT compound
  }

  override def readFromNBT(compound: NBTTagCompound) {
    super.readFromNBT(compound)
    mask(0) = compound getInteger "MaskId"
    mask(1) = compound getInteger "MaskMeta"
    if (compound hasKey "WTMultiblock") {
      new MultiblockWallTeleporter() add this
      multiblock readFromNBT compound
    }
  }

  override def updateEntity() {
    if (!_loaded && hasWorldObj) {
      _loaded = true
      if (!worldObj.isRemote && multiblock != null) multiblock init worldObj
      else if (worldObj.isRemote) {
        notifyNeighboursOfConnectionChanged()
        _startCountdown = true
      }
    }
    if (_countdown != -1 && _startCountdown) _countdown -= 1
    if (_countdown == 0 && multiblock.isController(this)) PacketHandler.requestMultiblockInfoPacket(this)

    if (cooldown) {
      multiblock.cooldown -= 1
      if (multiblock.cooldown == -1) {
        cooldown = false
      }
    }
  }

  override def validate() {
    super.validate()
    if (FMLCommonHandler.instance.getEffectiveSide.isClient) PacketHandler.requestMaskPacket(0.toByte, this)
  }

  override def invalidate() {
    super.invalidate()
    notifyNeighboursOfConnectionChanged()
  }

  override def getSizeInventory: Int = multiblock.getSizeInventory

  override def getStackInSlot(i: Int): ItemStack = multiblock getStackInSlot i

  override def decrStackSize(i: Int, j: Int): ItemStack = multiblock.decrStackSize(i, j)

  override def getStackInSlotOnClosing(i: Int): ItemStack = multiblock.getStackInSlotOnClosing(i)

  override def setInventorySlotContents(i: Int, stack: ItemStack) {
    multiblock.setInventorySlotContents(i, stack)
  }

  override def getInvName: String = multiblock.getInvName

  override def isInvNameLocalized: Boolean = multiblock.isInvNameLocalized

  override def getInventoryStackLimit: Int = multiblock.getInventoryStackLimit

  override def isUseableByPlayer(player: EntityPlayer): Boolean = multiblock.isUseableByPlayer(player)

  override def openChest() {}

  override def closeChest() {}

  override def isItemValidForSlot(i: Int, stack: ItemStack): Boolean = multiblock.isItemValidForSlot(i, stack)

  override def receiveIntPacket(id: Byte, integer: Int) {
    multiblock.receiveIntPacket(id, integer)
  }

  private var cooldown: Boolean = false

  def startCooldown() {
    multiblock.cooldown = 10
    cooldown = true
  }


  ////
  // Computer Craft Integration
  ////

  def getType: String = "wallTeleporter"

  def getMethodNames: Array[String] = Array("getX", "getY", "getZ", "getRotation", "getWorldId", "getWorldName", "clear",
                                            "hasDestination", "setUseRotation", "getUseRotation", "setMaskLocked",
                                            "isMaskLocked", "setMask", "getMask", "getFuelLevel")

  def callMethod(computer: IComputerAccess, context: ILuaContext, method: Int, arguments: Array[AnyRef]): Array[AnyRef] = {
    method match {
      case 0 => CCHelpers.argsOfOne(multiblock.destinationX)
      case 1 => CCHelpers.argsOfOne(multiblock.destinationY)
      case 2 => CCHelpers.argsOfOne(multiblock.destinationZ)
      case 3 => CCHelpers.argsOfOne(multiblock.destinationRotation)
      case 4 => CCHelpers.argsOfOne(multiblock.destinationWorldId)
      case 5 => CCHelpers.argsOfOne(multiblock.destinationWorldName)
      case 6 => multiblock.clearDestination(); CCHelpers.nullArgs
      case 7 => CCHelpers.argsOfOne(multiblock.hasDestination)
      case 8 =>
        if (!arguments(0).isInstanceOf[Boolean])  throw new Exception("Argument 1 expected to be of type boolean")
        multiblock.useRotation = arguments(0).asInstanceOf[Boolean]
        CCHelpers.nullArgs
      case 9 => CCHelpers.argsOfOne(multiblock.useRotation)
      case 10 =>
        if (!arguments(0).isInstanceOf[Boolean])  throw new Exception("Argument 1 expected to be of type boolean")
        multiblock.locked = arguments(0).asInstanceOf[Boolean]
        CCHelpers.nullArgs
      case 11 => CCHelpers.argsOfOne(multiblock.locked)
      case 12 =>
        if (!arguments(0).isInstanceOf[Double])  throw new Exception("Argument 1 expected to be of type number")
        if (!arguments(1).isInstanceOf[Double])  throw new Exception("Argument 2 expected to be of type number")
        setMask(arguments(0).asInstanceOf[Double].toInt, arguments(1).asInstanceOf[Double].toInt)
        CCHelpers.nullArgs
      case 13 => CCHelpers.argsOfTwo(mask(0), mask(1))
      case 14 => CCHelpers.argsOfOne(multiblock.getTrips)
      case _ => CCHelpers.nullArgs
    }
  }

  def canAttachToSide(side: Int): Boolean = true

  def attach(computer: IComputerAccess) {}

  def detach(computer: IComputerAccess) {}
}
